lib/SILOptimizer/Utils/PartialApplyCombiner.cpp

```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
 
-    for (auto copy : tmpCopies) {
-      builder.setInsertionPoint(term);
-      builder.createDeallocStack(pai->getLoc(), copy);
+    SILValue tmp = arg;
+    if (arg->getType().isObject()) {
+      tmp = builder.emitCopyValueOperation(pai->getLoc(), arg);
+    } else {
+      // Copy address-arguments into a stack-allocated temporary.
+      tmp = builder.createAllocStack(pai->getLoc(), arg->getType(), dbgVar);
+      builder.createCopyAddr(pai->getLoc(), arg, tmp, IsTake_t::IsNotTake,
+                             IsInitialization_t::IsInitialization);
     }
-  }
-}
+    argToTmpCopy.insert(std::make_pair(arg, tmp));
 
-/// Emit code to release/destroy temporaries.
-void PartialApplyCombiner::destroyTemporaries() {
-  // Insert releases and destroy_addrs as early as possible,
-  // because we don't want to keep objects alive longer than
-  // its really needed.
-  for (auto op : tmpCopies) {
-    auto tmpType = op->getType().getObjectType();
-    if (tmpType.isTrivial(*pai->getFunction()))
-      continue;
-    for (auto *endPoint : partialApplyFrontier) {
-      builder.setInsertionPoint(endPoint);
-      if (!tmpType.isAddressOnly(*pai->getFunction())) {
-        SILValue load = builder.emitLoadValueOperation(
-            pai->getLoc(), op, LoadOwnershipQualifier::Take);
-        builder.emitDestroyValueOperation(pai->getLoc(), load);
-      } else {
-        builder.createDestroyAddr(pai->getLoc(), op);
-      }
-    }
+    // Destroy the argument value (either as SSA value or in the stack-
+    // allocated temporary) at the end of the partial_apply's lifetime.
+    endLifetimeAtFrontier(tmp, partialApplyFrontier, builder);
   }
+  return true;
 }
 
 /// Process an apply instruction which uses a partial_apply
 /// as its callee.
 /// Returns true on success.
-bool PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
-  builder.setInsertionPoint(paiAI.getInstruction());
-  builder.setCurrentDebugScope(paiAI.getDebugScope());
-
-  // Prepare the args.
+void PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
+  // The arguments of the final apply instruction.
   SmallVector<SILValue, 8> argList;
-  // First the ApplyInst args.
+  // First, add the arguments of ther original ApplyInst args.
   for (auto Op : paiAI.getArguments())
     argList.push_back(Op);
 
-  SILInstruction *insertPoint = &*builder.getInsertionPoint();
-  // Next, the partial apply args.
-
-  // Pre-process partial_apply arguments only once, lazily.
-  if (isFirstTime) {
-    isFirstTime = false;
-    if (!allocateTemporaries())
-      return false;
-  }
-
-  // Now, copy over the partial apply args.
-  for (auto arg : pai->getArguments()) {
-    // If there is new temporary for this argument, use it instead.
-    if (argToTmpCopy.count(arg)) {
-      arg = argToTmpCopy.lookup(arg);
-    }
-    argList.push_back(arg);
-  }
-
-  builder.setInsertionPoint(insertPoint);
+  builder.setInsertionPoint(paiAI.getInstruction());
```

Comment:
Please change this to use a new SILBuilderWithScope that takes in a SILBuilderContext.